File: C++/1038.cpp
__________________________________________________________________________________________________
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    
    TreeNode* bstToGst(TreeNode* root) {
        setSum(root);
        return root;
    }
    
    int setSum(TreeNode* x, int delta = 0) {
        int ret = x->val;
        x->val += delta;
        if (x->right) {
            int right = setSum(x->right, delta);
            ret += right;
            x->val += right;
        }
        if (x->left) ret += setSum(x->left, ret + delta);
        return ret;
    }
};

__________________________________________________________________________________________________

__________________________________________________________________________________________________

File: Python3/1038.py
__________________________________________________________________________________________________
sample 20 ms submission
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def bstToGst(self, root: TreeNode) -> TreeNode:
        self.recBstToGst(root, 0)
        return root
        
    def recBstToGst(self, root: TreeNode, preVal: int) -> int:
        if root is None:
            return preVal
        
        res = self.recBstToGst(root.right, preVal)
        root.val += res
        return self.recBstToGst(root.left, root.val)
__________________________________________________________________________________________________
sample 24 ms submission
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def bstToGst(self, root: TreeNode) -> TreeNode:
        def recur(node, sumTillNow):
            if not node:
                return sumTillNow
            node.val += recur(node.right, sumTillNow)
            return recur(node.left, node.val)
            
        recur(root, 0)
        return root
__________________________________________________________________________________________________
